### PR TITLE
feat: parallelize downloads from TS compiler

### LINKED ISSUE #2626: TS compiler loads modules synchronously
This issue came up when working on rewrite of DenoDir and is a blocker for v1.0 (#2473)

In the current setup TS compiler requests module synchronously one-by-one. 
Modules are requested in two separate (synchronous) functions:
- `resolveModuleNames` - which is called for all imports in given TS file, compiler wants to know fully resolved URL to that module, but we're actually downloading the module in this step by issuing `op_fetch_module_meta_data` (synchronous).
- `getSourceFile` - this function returns source code of module. Since `op_fetch_module_meta_data` was already called in `resolveModuleNames` the files are already downloaded.

Ideas for solutions:
- During `resolveModuleNames` I wanted to send multiple specifiers to Rust to kick off parallel download of modules and block on it, but @ry suggested it's not possible since it's sync op. 
- During `resolveModuleNames` send one message to Rust for each specifier kicking off download of that module in background - this could be represented as `ModuleMetaDataFuture` that is stored in `DenoDir`. During subsequent `getSourceFile` call that future should be resolved. Although I'm not sure if it doesn't present the same problem as above - future won't be polled because it's a sync op.
- Explore TypeScript compiler API to find async alternatives to `resolveModuleNames` and `getSourceFile`

CC @ry @kitsonk

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes #2626

A small change to the compiler that requests all imports from a file at once instead of doing N round trips. Still needs a bit of cleanup.

`master`
```
$ time deno -r fetch https://deno.land/std/testing/runner.ts
Download https://deno.land/std/testing/runner.ts
Compile https://deno.land/std/testing/runner.ts
Download https://deno.land/std/flags/mod.ts
Download https://deno.land/std/fs/mod.ts
Download https://deno.land/std/testing/mod.ts
Download https://deno.land/std/fs/empty_dir.ts
Download https://deno.land/std/fs/ensure_dir.ts
Download https://deno.land/std/fs/ensure_file.ts
Download https://deno.land/std/fs/ensure_link.ts
Download https://deno.land/std/fs/ensure_symlink.ts
Download https://deno.land/std/fs/exists.ts
Download https://deno.land/std/fs/glob.ts
Download https://deno.land/std/fs/globrex.ts
Download https://deno.land/std/fs/move.ts
Download https://deno.land/std/fs/copy.ts
Download https://deno.land/std/fs/read_file_str.ts
Download https://deno.land/std/fs/write_file_str.ts
Download https://deno.land/std/fs/read_json.ts
Download https://deno.land/std/fs/write_json.ts
Download https://deno.land/std/fs/walk.ts
Download https://deno.land/std/fs/eol.ts
Download https://deno.land/std/fs/utils.ts
Download https://deno.land/std/fs/path/mod.ts
Download https://deno.land/std/fs/path/win32.ts
Download https://deno.land/std/fs/path/posix.ts
Download https://deno.land/std/fs/path/constants.ts
Download https://deno.land/std/fs/path/interface.ts
Download https://deno.land/std/fs/path/utils.ts
Download https://deno.land/std/testing/asserts.ts
Download https://deno.land/std/fmt/colors.ts
Download https://deno.land/std/testing/diff.ts
Download https://deno.land/std/testing/format.ts

real	0m4.273s
user	0m2.615s
sys	0m0.310s
```

`this PR`
```
$ time target/release/deno -r fetch https://deno.land/std/testing/runner.ts
Download https://deno.land/std/testing/runner.ts
Compile https://deno.land/std/testing/runner.ts
Download https://deno.land/std/flags/mod.ts
Download https://deno.land/std/fs/mod.ts
Download https://deno.land/std/testing/mod.ts
Download https://deno.land/std/fs/empty_dir.ts
Download https://deno.land/std/fs/ensure_dir.ts
Download https://deno.land/std/fs/ensure_file.ts
Download https://deno.land/std/fs/ensure_link.ts
Download https://deno.land/std/fs/ensure_symlink.ts
Download https://deno.land/std/fs/exists.ts
Download https://deno.land/std/fs/glob.ts
Download https://deno.land/std/fs/globrex.ts
Download https://deno.land/std/fs/move.ts
Download https://deno.land/std/fs/copy.ts
Download https://deno.land/std/fs/read_file_str.ts
Download https://deno.land/std/fs/write_file_str.ts
Download https://deno.land/std/fs/read_json.ts
Download https://deno.land/std/fs/write_json.ts
Download https://deno.land/std/fs/walk.ts
Download https://deno.land/std/fs/eol.ts
Download https://deno.land/std/fs/utils.ts
Download https://deno.land/std/fs/path/mod.ts
Download https://deno.land/std/fs/path/win32.ts
Download https://deno.land/std/fs/path/posix.ts
Download https://deno.land/std/fs/path/constants.ts
Download https://deno.land/std/fs/path/interface.ts
Download https://deno.land/std/fs/path/utils.ts
Download https://deno.land/std/testing/asserts.ts
Download https://deno.land/std/fmt/colors.ts
Download https://deno.land/std/testing/diff.ts
Download https://deno.land/std/testing/format.ts

real	0m2.259s
user	0m2.494s
sys	0m0.250s
```

CC @ry @kitsonk

### CODE REVIEW

**Comment by bartlomieju** [comment]
I must admit, doing this PR with JSON ops is a breeze... I couldn't figure it out with Flatbufs a couple times.

File: cli/ops/compiler.rs [hunk 1]
```diff
@@ -33,39 +33,49 @@ pub fn op_cache(
...
+  // runtime for executing the future. This is so we don't inadvertently run
   // out of threads in the main runtime.
-  let out = tokio_util::block_on(fut)?;
-  Ok(JsonOp::Sync(json!({
-    "moduleName": out.url.to_string(),
-    "filename": out.filename.to_str().unwrap(),
-    "mediaType": out.media_type as i32,
-    "sourceCode": String::from_utf8(out.source_code).unwrap(),
-  })))
+  let files = tokio_util::block_on(futures::future::join_all(futures))?;
```
Comment (ry) [inline cli/ops/compiler.rs]: 👍 elegant

(at some point we need to get rid of this block_on stuff tho...)

**Comment by ry** [comment]
This caused the thread count to jump from 11 to 77 threads in cold_relative_import:

Screen Shot 2019-09-15 at 10 38 43 AM

This is due to parallel usage of block_on... :(

**Comment by bartlomieju** [comment]
This is bad news. We really need to figure out what to do with `block_on`. Wouldn't small increase of default number of threads solve this issue?

**Comment by ry** [comment]
@bartlomieju The issue is that we're starting a whole new tokio runtime for each module resolved - it is wildly inefficient. 

> Wouldn't small increase of default number of threads solve this issue?

No

**Comment by ry** [comment]
@bartlomieju After some thought, I will not revert your patch. Your patch speeds up an important (unfortunately unmeasured) benchmark. I think what you're doing is more or less correct and just exposing a lower-level problem. I've added https://github.com/denoland/deno/issues/2960 to 1.0 blockers to solve this most important design problem.

**Comment by bartlomieju** [comment]
> @bartlomieju After some thought, I will not revert your patch. Your patch speeds up an important (unfortunately unmeasured) benchmark. I think what you're doing is more or less correct and just exposing a lower-level problem. I've added #2960 to 1.0 blockers to solve this most important design problem.

@ry alright, I will investigate alternate solutions as well

**Pull request merged**